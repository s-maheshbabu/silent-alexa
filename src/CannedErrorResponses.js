const { Map } = require("immutable");

/**
 * We need to show human readable error messages to users when Alexa responds
 * with error codes or we encounter internal/unknown errors.
 * This is a mapping from Alexa/Custom error codes to canned responses.
 * @link https://developer.amazon.com/docs/alexa-voice-service/exceptions.html
 */
// TODO: Change the place holder error responses to real ones.
const cannedErrorResponses = Map({
  INVALID_REQUEST_EXCEPTION: "a canned response for INVALID_REQUEST_EXCEPTION",
  UNAUTHORIZED_REQUEST_EXCEPTION:
    "a canned response for UNAUTHORIZED_REQUEST_EXCEPTION",
  UNSUPPORTED_MEDIA_TYPE: "a canned response for UNSUPPORTED_MEDIA_TYPE",
  THROTTLING_EXCEPTION: "a canned response for THROTTLING_EXCEPTION",
  INTERNAL_SERVICE_EXCEPTION:
    "a canned response for INTERNAL_SERVICE_EXCEPTION",
  "N/A": "a canned response for N/A Exception",

  UNKNOWN_ERROR: "a canned response for UNKNOWN_ERROR"
});

const customErrorCodes = Object.freeze({
  UNKNOWN_ERROR: "UNKNOWN_ERROR"
});

module.exports = {
  cannedErrorResponses: cannedErrorResponses,
  customErrorCodes: customErrorCodes
};
