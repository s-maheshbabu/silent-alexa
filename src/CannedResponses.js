const cannedResponses = Object.freeze({
  /* Message to be delivered when Alexa responds with an empty message. This is a valid scenario,
   for example, when user says "shutup" to Alexa.*/
  EMPTY_RESPONSE_FROM_ALEXA:
    "a canned response when Alexa repsonds with an empty message."
});

module.exports = {
  cannedResponses: cannedResponses
};
