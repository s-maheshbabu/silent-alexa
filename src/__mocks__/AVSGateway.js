export const mockAlexaSuccessResponse = "mock alexa success response";

// Mock implementation that always resolves to a successful response from AVS.
export const mockSendTextMessageEventFunction = jest.fn(() =>
  Promise.resolve(mockAlexaSuccessResponse)
);

const mockAVSGateway = jest.fn().mockImplementation(() => {
  return { sendTextMessageEvent: mockSendTextMessageEventFunction };
});

export default mockAVSGateway;
