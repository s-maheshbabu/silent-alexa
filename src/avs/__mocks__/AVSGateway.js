import { List } from "immutable";

export const mockAlexaSuccessResponses = List.of(
  "mock alexa success response 1",
  "mock alexa success response 2",
  "mock alexa success response 3"
);

// Mock implementation that always resolves to a successful response from AVS.
export const mockSendTextMessageEventFunction = jest.fn(() =>
  Promise.resolve(mockAlexaSuccessResponses)
);

// Mock implementation that always resolves to a successful response from AVS.
export const mockSendAddOrUpdateReportEventFunction = jest.fn(() => Promise.resolve(true)
);

const mockAVSGateway = jest.fn(() => {
  return {
    sendTextMessageEvent: mockSendTextMessageEventFunction,
    sendAddOrUpdateReportEvent: mockSendAddOrUpdateReportEventFunction
  };
});

export default mockAVSGateway;
