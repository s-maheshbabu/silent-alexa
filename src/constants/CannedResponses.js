const cannedResponses = Object.freeze({
  /* Message to be delivered when Alexa responds with an empty message. This is a valid scenario,
   for example, when user says "stop talking" to Alexa.*/
  EMPTY_RESPONSE_FROM_ALEXA:
    "a canned response when Alexa repsonds with an empty message."
});

module.exports = {
  cannedResponses: cannedResponses
};
