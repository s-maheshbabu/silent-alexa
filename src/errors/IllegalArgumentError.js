class IllegalArgumentError extends Error {
  constructor(message) {
    super(message);
    this.message = message;
    this.name = "IllegalArgumentError";
  }
}

export default IllegalArgumentError;
