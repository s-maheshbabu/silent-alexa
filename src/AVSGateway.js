import { urls, paths } from "./AVSEndPoints";
import { cannedErrorResponses, customErrorCodes } from "./CannedErrorResponses";
import { cannedResponses } from "./CannedResponses";

import IllegalArgumentError from "./errors/IllegalArgumentError";
import { extractAlexaTextResponses as parser } from "./SpeakDirectiveParser";

import { hasIn, List } from "immutable";
import uuid from "uuid/v4";
import util from "util";

const sprintf = require("sprintf-js").sprintf;

const AVS_REQUEST_BODY = `--silent-alexa-http-boundary
Content-Disposition: form-data; name="metadata"
Content-Type: application/json; charset=UTF-8

%s`;

// TODO: Once the region setting is made configurable by the user, the URLS
// need to ge generated as against hard coded.
export const EVENTS_URL = urls.NA + paths.EVENTS;

/**
 * Manages interactions with AVS
 */
export default class AVSGateway {
  /**
   * Sends the TextMessage event to AVS and extracts Alexa's responses.
   *
   * @param {String} userRequestToAlexa The request string that the user typed
   * as a request for Alexa. This should not be empty or undefined.
   * @param {String} accessToken The access token to communicate with AVS. This
   * should not be empty of undefined.
   *
   * @returns A list of text responses from Alexa. Will never return undefined or
   * empty list. If an error happens while communicating to AVS or while parsing
   * the responses, a canned human-readable error message is returned.
   *
   * @throws IllegalArgumentError if the input is missing or invalid.
   */
  async sendTextMessageEvent(userRequestToAlexa, accessToken) {
    if (!userRequestToAlexa || !accessToken) {
      throw new IllegalArgumentError(
        `The request string to Alexa or the access token cannot be empty.
        RequestString: ${userRequestToAlexa}
        AccessToken: ${accessToken}`
      );
    }

    const testMessageEvent = JSON.stringify(
      this.buildTextMessageEvent(userRequestToAlexa)
    );
    const requestOptions = this.buildTextMessageFetchRequestOptions(
      testMessageEvent,
      accessToken
    );

    let isOk = false;
    let payload;
    await fetch(EVENTS_URL, requestOptions)
      .then(response => {
        if (response.ok) {
          isOk = true;
          return response.text();
        } else {
          return response.json();
        }
      })
      .then(data => {
        payload = data;
      })
      .catch(error => {
        isOk = false; // We obtained a successful response but couldn't parse the body, probably because it was malformed.
        console.log(util.inspect(error, { showHidden: true, depth: null }));
      });

    if (isOk) {
      try {
        let textResponsesFromAlexa;
        if (payload) textResponsesFromAlexa = parser(payload);
        if (
          !textResponsesFromAlexa ||
          // We don't anticipate Alexa to return a response in multiple parts where the first part is empty
          // but the other parts aren't. So, the moment we see that the first part is empty, it is safe to
          // ignore all other parts (which probably don't exist).
          !textResponsesFromAlexa.get(0)
        )
          textResponsesFromAlexa = List.of(
            cannedResponses.EMPTY_RESPONSE_FROM_ALEXA
          );

        return textResponsesFromAlexa;
      } catch (error) {
        console.log(
          "Encountered an error while trying to parse the speak directive from AVS." +
            util.inspect(error, { showHidden: true, depth: null })
        );
      }
    }

    return List.of(this.convertErrorToHumanReadableMessage(payload));
  }

  convertErrorToHumanReadableMessage(errorPayload) {
    let errorCode;
    if (hasIn(errorPayload, ["payload", "code"]))
      errorCode = errorPayload.payload.code;

    return cannedErrorResponses.get(
      errorCode,
      // default value to return if errorCode doesn't exist
      cannedErrorResponses.get(customErrorCodes.UNKNOWN_ERROR)
    );
  }

  buildTextMessageFetchRequestOptions(testMessageEvent, accessToken) {
    const data = sprintf(AVS_REQUEST_BODY, testMessageEvent);

    return {
      body: data,
      headers: {
        Authorization: "Bearer " + accessToken,
        "content-type":
          "multipart/form-data; boundary=silent-alexa-http-boundary"
      },
      cache: "no-store", // Alexa often responds differently to the same request and so we don't want to cache anything.
      method: "POST"
    };
  }

  buildTextMessageEvent(requestString) {
    return {
      event: {
        header: {
          namespace: "Text",
          name: "TextMessage",
          messageId: uuid()
        },
        payload: {
          textMessage: requestString
        }
      }
    };
  }
}
