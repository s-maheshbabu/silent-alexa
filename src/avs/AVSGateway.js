import { urls, paths } from "AVSEndPoints";
import { cannedErrorResponses, customErrorCodes } from "CannedErrorResponses";
import { cannedResponses } from "CannedResponses";

import IllegalArgumentError from "errors/IllegalArgumentError";
import { extractAlexaTextResponses as parser } from "SpeakDirectiveParser";

import { hasIn, List } from "immutable";
import uuid from "uuid/v4";
import util from "util";

import { PRODUCT_ID, CLIENT_ID } from "Constants";

const sprintf = require("sprintf-js").sprintf;

const AVS_REQUEST_BODY = `--silent-alexa-http-boundary
Content-Disposition: form-data; name="metadata"
Content-Type: application/json; charset=UTF-8

%s`;

// TODO: Once the region setting is made configurable by the user, the URLS
// need to ge generated as against hard coded.
export const EVENTS_URL = urls.NA + paths.EVENTS;

/**
 * Manages interactions with AVS
 */
export default class AVSGateway {
  /**
   * Sends the TextMessage event to AVS and extracts Alexa's responses.
   *
   * @param {String} userRequestToAlexa The request string that the user typed
   * as a request for Alexa. This should not be empty or undefined.
   * @param {String} accessToken The access token to communicate with AVS. This
   * should not be empty of undefined.
   *
   * @returns A list of text responses from Alexa. Will never return undefined or
   * empty list. If an error happens while communicating to AVS or while parsing
   * the responses, a canned human-readable error message is returned.
   *
   * @throws IllegalArgumentError if the input is missing or invalid.
   */
  async sendTextMessageEvent(userRequestToAlexa, accessToken) {
    if (!userRequestToAlexa || !accessToken) {
      throw new IllegalArgumentError(
        `The request string to Alexa or the access token cannot be empty.
        RequestString: ${userRequestToAlexa}
        AccessToken: ${accessToken}`
      );
    }

    const textMessageEvent = JSON.stringify(
      this.buildTextMessageEvent(userRequestToAlexa)
    );
    const requestOptions = this.buildTextMessageFetchRequestOptions(
      textMessageEvent,
      accessToken
    );

    let isOk = false;
    let payload;
    await fetch(EVENTS_URL, requestOptions)
      .then(response => {
        if (response.ok) {
          isOk = true;
          return response.text();
        } else {
          return response.json();
        }
      })
      .then(data => {
        payload = data;
      })
      .catch(error => {
        isOk = false; // We obtained a successful response but couldn't parse the body, probably because it was malformed.
        console.log(util.inspect(error, { showHidden: true, depth: null }));
      });

    if (isOk) {
      try {
        let textResponsesFromAlexa;
        if (payload) textResponsesFromAlexa = parser(payload);
        if (
          !textResponsesFromAlexa ||
          // We don't anticipate Alexa to return a response in multiple parts where the first part is empty
          // but the other parts aren't. So, the moment we see that the first part is empty, it is safe to
          // ignore all other parts (which probably don't exist).
          !textResponsesFromAlexa.get(0)
        )
          textResponsesFromAlexa = List.of(
            cannedResponses.EMPTY_RESPONSE_FROM_ALEXA
          );

        return textResponsesFromAlexa;
      } catch (error) {
        console.log(
          "Encountered an error while trying to parse the speak directive from AVS." +
          util.inspect(error, { showHidden: true, depth: null })
        );
      }
    }

    return List.of(this.convertErrorToHumanReadableMessage(payload));
  }

  /**
   * Sends the AddOrUpdateReport event to AVS. This event is meant to proactively
   * inform Alexa about the endpoint.
   * 
   * @returns true if the event has been successfully processed by Alexa and
   * false otherwise. Alexa requires that no other events should be sent if this
   * event is not processed successfully.
   */
  // TODO: We should probably retry three times before propagating failures.
  async sendAddOrUpdateReportEvent(accessToken) {
    const addOrUpdateReportEvent = JSON.stringify(
      this.buildAddOrUpdateReportEvent(accessToken)
    );
    const requestOptions = this.buildAddOrUpdateReportEventFetchRequestOptions(
      addOrUpdateReportEvent,
      accessToken
    );

    let isOk = false;
    await fetch(EVENTS_URL, requestOptions)
      .then(response => {
        if (response.ok) {
          isOk = true;
        }
      }).catch(error => {
        console.log(util.inspect(error, { showHidden: true, depth: null }));
      });;

    return isOk;
  }

  convertErrorToHumanReadableMessage(errorPayload) {
    let errorCode;
    if (hasIn(errorPayload, ["payload", "code"]))
      errorCode = errorPayload.payload.code;

    return cannedErrorResponses.get(
      errorCode,
      // default value to return if errorCode doesn't exist
      cannedErrorResponses.get(customErrorCodes.UNKNOWN_ERROR)
    );
  }

  buildTextMessageFetchRequestOptions(textMessageEvent, accessToken) {
    const data = sprintf(AVS_REQUEST_BODY, textMessageEvent);

    return {
      body: data,
      headers: {
        Authorization: "Bearer " + accessToken,
        "content-type":
          "multipart/form-data; boundary=silent-alexa-http-boundary"
      },
      cache: "no-store", // Alexa often responds differently to the same request and so we don't want to cache anything.
      method: "POST"
    };
  }

  buildTextMessageEvent(requestString) {
    return {
      event: {
        header: {
          namespace: "Text",
          name: "TextMessage",
          messageId: uuid()
        },
        payload: {
          textMessage: requestString
        }
      }
    };
  }

  buildAddOrUpdateReportEventFetchRequestOptions(addOrUpdateReportEvent, accessToken) {
    const data = sprintf(AVS_REQUEST_BODY, addOrUpdateReportEvent);

    return {
      body: data,
      headers: {
        Authorization: "Bearer " + accessToken,
        "content-type":
          "multipart/form-data; boundary=silent-alexa-http-boundary"
      },
      cache: "no-store", // Alexa often responds differently to the same request and so we don't want to cache anything.
      method: "POST"
    };
  }

  buildAddOrUpdateReportEvent(accessToken) {
    // TODO: Logic for assigning 'deviceSerialNumber' needs to be revisited.
    const DSN = "12345";

    return {
      event: {
        header: {
          namespace: "Alexa.Discovery",
          name: "AddOrUpdateReport",
          payloadVersion: "3",
          messageId: uuid(),
          eventCorrelationToken: uuid()
        },
        payload: {
          scope: {
            type: "BearerToken",
            token: accessToken
          },
          endpoints: [
            {
              endpointId: `${CLIENT_ID}::${PRODUCT_ID}::${DSN}`,
              registration: {
                productId: PRODUCT_ID,
                deviceSerialNumber: DSN
              },
              manufacturerName: "Silent Voice Assistants",
              description: "Interact with voice assistants without having to talk to them.",
              friendlyName: "Silent Alexa",
              displayCategories: ["COMPUTER", "LAPTOP", "TABLET"],
              capabilities: [
                {
                  type: "AlexaInterface",
                  interface: "SpeechSynthesizer",
                  version: "1.0"
                }
              ],
              connections: [
                {
                  type: "UNKNOWN",
                  value: DSN //TODO: Is it reasonable to use a DSN here?
                }
              ]
            }
          ]
        }
      }
    };
  }
}
