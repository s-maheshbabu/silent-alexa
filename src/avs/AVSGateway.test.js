import { List } from "immutable";

import AVSGateway, { EVENTS_URL } from "./AVSGateway";
import IllegalArgumentError from "errors/IllegalArgumentError";

import { cannedErrorResponses, customErrorCodes } from "CannedErrorResponses";
import { cannedResponses } from "CannedResponses";

import testData from "./test-data/multipart-response-test-data";
import addOrUpdateReportTestData from "./test-data/EventProcessedDirective";

// A mock parser that uses the real implementation by default.
jest.mock("SpeakDirectiveParser", () => {
  return {
    extractAlexaTextResponses: jest.fn(
      require.requireActual("SpeakDirectiveParser").extractAlexaTextResponses
    )
  };
});
const parser = require("SpeakDirectiveParser");

jest.mock("uuid/v4", () => {
  return jest.fn(() => "mock-uuid-generated-for-testing");
});

import fetchMock from "fetch-mock";

const unitUnderTest = new AVSGateway();

afterEach(() => {
  fetchMock.restore();
});

test("that sending a textRequest throws if the message itself is null, empty or undefined.", async () => {
  expect.assertions(3);

  const textMessage = "";
  await expect(unitUnderTest.sendTextMessageEvent(textMessage)).rejects.toEqual(
    expect.any(IllegalArgumentError)
  );

  let nullMessage;
  await expect(unitUnderTest.sendTextMessageEvent(nullMessage)).rejects.toEqual(
    expect.any(IllegalArgumentError)
  );

  await expect(unitUnderTest.sendTextMessageEvent(undefined)).rejects.toEqual(
    expect.any(IllegalArgumentError)
  );
});

test("that sending a textRequest throws if the accessToken is null, empty or undefined.", async () => {
  expect.assertions(3);

  const userRequestToAlexa = "user request to Alexa";
  const accessToken = "";

  await expect(
    unitUnderTest.sendTextMessageEvent(userRequestToAlexa, accessToken)
  ).rejects.toEqual(expect.any(IllegalArgumentError));

  let nullAccessToken;
  await expect(
    unitUnderTest.sendTextMessageEvent(userRequestToAlexa, nullAccessToken)
  ).rejects.toEqual(expect.any(IllegalArgumentError));

  await expect(
    unitUnderTest.sendTextMessageEvent(userRequestToAlexa, undefined)
  ).rejects.toEqual(expect.any(IllegalArgumentError));
});

it("handles gracefully if AVS returns INVALID_REQUEST_EXCEPTION error code", () => {
  testErrorResponseFromAVSHandling("INVALID_REQUEST_EXCEPTION", 400);
});

it("handles gracefully if AVS returns UNAUTHORIZED_REQUEST_EXCEPTION error code", () => {
  testErrorResponseFromAVSHandling("UNAUTHORIZED_REQUEST_EXCEPTION", 403);
});

it("handles gracefully if AVS returns UNSUPPORTED_MEDIA_TYPE error code", () => {
  testErrorResponseFromAVSHandling("UNSUPPORTED_MEDIA_TYPE", 415);
});

it("handles gracefully if AVS returns THROTTLING_EXCEPTION error code", () => {
  testErrorResponseFromAVSHandling("THROTTLING_EXCEPTION", 429);
});

it("handles gracefully if AVS returns INTERNAL_SERVICE_EXCEPTION error code", () => {
  testErrorResponseFromAVSHandling("INTERNAL_SERVICE_EXCEPTION", 500);
});

it("handles gracefully if AVS returns N/A error code", () => {
  testErrorResponseFromAVSHandling("N/A", 503);
});

it("handles gracefully if AVS returns an unexpected error code", async () => {
  const userRequestToAlexa = "user request to Alexa";
  const access_token = "a mock access_token";
  const errorCode = "an unexpected error code";

  fetchMock.postOnce(EVENTS_URL, {
    status: 500,
    body: {
      header: {
        namespace: "System",
        name: "Exception",
        messageId: "a message id"
      },

      payload: {
        code: errorCode,
        description: "description of the error"
      }
    }
  });

  await unitUnderTest
    .sendTextMessageEvent(userRequestToAlexa, access_token)
    .then(alexaResponse => {
      expect(alexaResponse).toEqual(
        List.of(cannedErrorResponses.get(customErrorCodes.UNKNOWN_ERROR))
      );
    });
});

it("handles gracefully when the http response is 'ok' and then we fail to parse the text body out of it.", async () => {
  const userRequestToAlexa = "user request to Alexa";
  const access_token = "a mock access_token";

  fetchMock.postOnce(EVENTS_URL, {
    status: 200,
    body: undefined // response.text() will parse pretty much anything you throw at it. The only way I found to make it fail is to pass it undefined as input.
  });

  await unitUnderTest
    .sendTextMessageEvent(userRequestToAlexa, access_token)
    .then(alexaResponse => {
      expect(alexaResponse).toEqual(
        List.of(cannedErrorResponses.get(customErrorCodes.UNKNOWN_ERROR))
      );
    });
});

it("handles gracefully when the speak directive parser throws an error", async () => {
  const userRequestToAlexa = "user request to Alexa";
  const access_token = "a mock access_token";

  // Mock the parser to throw an illegal argument error
  parser.extractAlexaTextResponses.mockImplementationOnce(() => {
    throw new IllegalArgumentError("simulating an illegal argument error");
  });

  fetchMock.postOnce(EVENTS_URL, {
    status: 200,
    body: testData.happy_case.rawData
  });

  await unitUnderTest
    .sendTextMessageEvent(userRequestToAlexa, access_token)
    .then(alexaResponse => {
      expect(alexaResponse).toEqual(
        List.of(cannedErrorResponses.get(customErrorCodes.UNKNOWN_ERROR))
      );
    });
});

it("handles gracefully when the http response is not 'ok' and then we fail to parse the json body out of it.", async () => {
  const userRequestToAlexa = "user request to Alexa";
  const access_token = "a mock access_token";

  fetchMock.postOnce(EVENTS_URL, {
    status: 500,
    body: "invalid json string"
  });

  await unitUnderTest
    .sendTextMessageEvent(userRequestToAlexa, access_token)
    .then(alexaResponse => {
      expect(alexaResponse).toEqual(
        List.of(cannedErrorResponses.get(customErrorCodes.UNKNOWN_ERROR))
      );
    });
});

it("calls fetch with the right options when trying to send a happy case TextMessage event and extracts the response string", async () => {
  const userRequestToAlexa = "user request to Alexa";
  const access_token = "a mock access_token";

  fetchMock.postOnce(EVENTS_URL, testData.happy_case.rawData);

  await unitUnderTest
    .sendTextMessageEvent(userRequestToAlexa, access_token)
    .then(alexaResponse => {
      expect(alexaResponse).toEqual(testData.happy_case.alexaResponses);
    });

  const requestOptionsUsed = fetchMock.lastOptions();
  expect(requestOptionsUsed).toMatchSnapshot();
});

it(`handles the happy case where Alexa returns an empty response (can happen when Alexa has nothing to say, for example, in response to user saying 'shutup')
or responds with nothing (can happen when user says something senseless like 'dramatic ink')`, async () => {
  const userRequestToAlexa = "user request to Alexa";
  const access_token = "a mock access_token";

  const emptyishResponses = [
    testData.happy_case_when_alexa_responds_with_empty_message.rawData,
    testData.happy_case_when_alexa_responds_with_nothing.rawData
  ];

  for (let emptyResponse of emptyishResponses) {
    fetchMock.post(
      EVENTS_URL,
      new Promise(function (resolve) {
        resolve(emptyResponse);
      }),
      {
        overwriteRoutes: true
      }
    );

    await unitUnderTest
      .sendTextMessageEvent(userRequestToAlexa, access_token)
      .then(alexaResponse => {
        expect(alexaResponse).toEqual(
          List.of(cannedResponses.EMPTY_RESPONSE_FROM_ALEXA)
        );
      });

    const requestOptionsUsed = fetchMock.lastOptions();
    expect(requestOptionsUsed).toMatchSnapshot();
  }
});

it("calls fetch with the right options and return true when trying to send a happy case AddOrUpdateReport event", async () => {
  const access_token = "a mock access_token";

  fetchMock.postOnce(EVENTS_URL, addOrUpdateReportTestData.happy_case);

  await unitUnderTest
    .sendAddOrUpdateReportEvent(access_token)
    .then(alexaResponse => {
      expect(alexaResponse).toBe(true);
    });

  const requestOptionsUsed = fetchMock.lastOptions();
  expect(requestOptionsUsed).toMatchSnapshot();
});

it("handles gracefully if AVS returns an error", async () => {
  const access_token = "a mock access_token";
  const error_codes = [400, 401, 403, 415, 429, 500, 503];

  for (let errorCode of error_codes) {
    fetchMock.postOnce(EVENTS_URL, {
      status: errorCode
    });

    await unitUnderTest
      .sendAddOrUpdateReportEvent(access_token)
      .then(alexaResponse => {
        expect(alexaResponse).toBe(false);
      });

    fetchMock.restore();
  }
});

/**
 * Verifies that the given input results in an {IllegalArgumentError} when parsed.
 * @param {String} input The multi part response string that needs to be parsed.
 */
const testErrorResponseFromAVSHandling = async (
  avsErrorCode,
  httpErrorCode
) => {
  const userRequestToAlexa = "user request to Alexa";
  const access_token = "a mock access_token";

  fetchMock.postOnce(EVENTS_URL, {
    status: httpErrorCode,
    body: {
      header: {
        namespace: "System",
        name: "Exception",
        messageId: "a message id"
      },

      payload: {
        code: avsErrorCode,
        description: "description of the error"
      }
    }
  });

  await unitUnderTest
    .sendTextMessageEvent(userRequestToAlexa, access_token)
    .then(alexaResponse => {
      expect(alexaResponse).toEqual(
        List.of(cannedErrorResponses.get(avsErrorCode))
      );
    });
};
