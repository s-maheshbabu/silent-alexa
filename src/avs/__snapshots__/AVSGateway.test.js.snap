// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calls fetch with the right options and return true when trying to send a happy case AddOrUpdateReport event 1`] = `
Object {
  "body": "--silent-alexa-http-boundary
Content-Disposition: form-data; name=\\"metadata\\"
Content-Type: application/json; charset=UTF-8

{\\"event\\":{\\"header\\":{\\"namespace\\":\\"Alexa.Discovery\\",\\"name\\":\\"AddOrUpdateReport\\",\\"payloadVersion\\":\\"3\\",\\"messageId\\":\\"mock-uuid-generated-for-testing\\",\\"eventCorrelationToken\\":\\"mock-uuid-generated-for-testing\\"},\\"payload\\":{\\"scope\\":{\\"type\\":\\"BearerToken\\",\\"token\\":\\"a mock access_token\\"},\\"endpoints\\":[{\\"endpointId\\":\\"amzn1.application-oa2-client.b7bd09596de34d37a2403a712f137e22::Silent_Alexa::12345\\",\\"registration\\":{\\"productId\\":\\"Silent_Alexa\\",\\"deviceSerialNumber\\":\\"12345\\"},\\"manufacturerName\\":\\"Silent Voice Assistants\\",\\"description\\":\\"Interact with voice assistants without having to talk to them.\\",\\"friendlyName\\":\\"Silent Alexa\\",\\"displayCategories\\":[\\"COMPUTER\\",\\"LAPTOP\\",\\"TABLET\\"],\\"capabilities\\":[{\\"type\\":\\"AlexaInterface\\",\\"interface\\":\\"SpeechSynthesizer\\",\\"version\\":\\"1.3\\"}],\\"connections\\":[{\\"type\\":\\"UNKNOWN\\",\\"value\\":\\"12345\\"}]}]}}}",
  "cache": "no-store",
  "headers": Object {
    "Authorization": "Bearer a mock access_token",
    "content-type": "multipart/form-data; boundary=silent-alexa-http-boundary",
  },
  "method": "POST",
}
`;

exports[`calls fetch with the right options when trying to send a happy case TextMessage event and extracts the response string 1`] = `
Object {
  "body": "--silent-alexa-http-boundary
Content-Disposition: form-data; name=\\"metadata\\"
Content-Type: application/json; charset=UTF-8

{\\"event\\":{\\"header\\":{\\"namespace\\":\\"Text\\",\\"name\\":\\"TextMessage\\",\\"messageId\\":\\"mock-uuid-generated-for-testing\\"},\\"payload\\":{\\"textMessage\\":\\"user request to Alexa\\"}}}",
  "cache": "no-store",
  "headers": Object {
    "Authorization": "Bearer a mock access_token",
    "content-type": "multipart/form-data; boundary=silent-alexa-http-boundary",
  },
  "method": "POST",
}
`;

exports[`handles the happy case where Alexa returns an empty response (can happen when Alexa has nothing to say, for example, in response to user saying 'shutup')
or responds with nothing (can happen when user says something senseless like 'dramatic ink') 1`] = `
Object {
  "body": "--silent-alexa-http-boundary
Content-Disposition: form-data; name=\\"metadata\\"
Content-Type: application/json; charset=UTF-8

{\\"event\\":{\\"header\\":{\\"namespace\\":\\"Text\\",\\"name\\":\\"TextMessage\\",\\"messageId\\":\\"mock-uuid-generated-for-testing\\"},\\"payload\\":{\\"textMessage\\":\\"user request to Alexa\\"}}}",
  "cache": "no-store",
  "headers": Object {
    "Authorization": "Bearer a mock access_token",
    "content-type": "multipart/form-data; boundary=silent-alexa-http-boundary",
  },
  "method": "POST",
}
`;

exports[`handles the happy case where Alexa returns an empty response (can happen when Alexa has nothing to say, for example, in response to user saying 'shutup')
or responds with nothing (can happen when user says something senseless like 'dramatic ink') 2`] = `
Object {
  "body": "--silent-alexa-http-boundary
Content-Disposition: form-data; name=\\"metadata\\"
Content-Type: application/json; charset=UTF-8

{\\"event\\":{\\"header\\":{\\"namespace\\":\\"Text\\",\\"name\\":\\"TextMessage\\",\\"messageId\\":\\"mock-uuid-generated-for-testing\\"},\\"payload\\":{\\"textMessage\\":\\"user request to Alexa\\"}}}",
  "cache": "no-store",
  "headers": Object {
    "Authorization": "Bearer a mock access_token",
    "content-type": "multipart/form-data; boundary=silent-alexa-http-boundary",
  },
  "method": "POST",
}
`;
